source/slang/emit.cpp

```diff
@@ -4129,7 +4135,98 @@ struct EmitVisitor
         }
     }
 
+    
+
+    IRInst* findFirstInst(IRFunc* irFunc, IROp op)
+    {
+        for (auto block : irFunc->getBlocks())
+        {
+            for (auto inst : block->getChildren())
+            {
+                if (inst->op == op)
+                {
+                    return inst;
+                }
+            }
+        }
+        return nullptr;
+    }
+
+    void emitAttributeSingleString(const char* name, FuncDecl* entryPoint, Attribute* attrib)
+    {
+        assert(attrib);
+        
+        attrib->args.Count();
+        if (attrib->args.Count() != 1)
+        {
+            SLANG_DIAGNOSE_UNEXPECTED(getSink(), entryPoint->loc, "Attribute expects single parameter");
+            return;
+        }
+
+        Expr* expr = attrib->args[0];
+
+        auto stringLitExpr = expr->As<StringLiteralExpr>();
+        if (!stringLitExpr)
+        {
+            SLANG_DIAGNOSE_UNEXPECTED(getSink(), entryPoint->loc, "Attribute parameter expecting to be a string ");
+            return;
+        }
+
+        emit("[");
+        emit(name);
+        emit("(\"");
+        emit(stringLitExpr->value); 
+        emit("\")]\n");
+    }
+
+    void emitAttributeSingleInt(const char* name, FuncDecl* entryPoint, Attribute* attrib)
+    {
+        assert(attrib);
+
+        attrib->args.Count();
+        if (attrib->args.Count() != 1)
+        {
+            SLANG_DIAGNOSE_UNEXPECTED(getSink(), entryPoint->loc, "Attribute expects single parameter");
+            return;
+        }
+
+        Expr* expr = attrib->args[0];
+
+        auto intLitExpr = expr->As<IntegerLiteralExpr>();
+        if (!intLitExpr)
+        {
+            SLANG_DIAGNOSE_UNEXPECTED(getSink(), entryPoint->loc, "Attribute expects an int");
+            return;
+        }
+
+        emit("[");
+        emit(name);
+        emit("(");
+        emit(intLitExpr->value);
```

Comment:
Seeing how much this shares with the `emitAttributeSingleString` function, I'm left wondering if we want a catch-all operation that emits an attribute with any number of string or integer-literal arguments.

But that just brings to the surface the more serious issue, which is that this is one of the main cases where the emit logic (which is nominally translating our IR over to HLSL/GLSL) is actually making heavy use of AST-level structures. This means that we haven't done a complete job of translating from AST->IR.

The "right" way to handle these would be for every one of the important attributes to translate over to either decorations or full instructions (like this PR does when turning `[patchconstantfunc(...)]` into a `notePatchConstantFunc(...)` instruction), and then emit entirely from that IR representation.

I suppose it would be better to invest in that "right way" of doing things rather than just spend a bunch more effort on cleaning up code that is doing things the "wrong way" using the AST.